content/en/docs/guides/performance.md

```diff
@@ -0,0 +1,102 @@
+    Applications with high load or long-lived streaming RPCs might see
+    performance issues because of this queueing. There are two possible
+    solutions:
+
+    1.  **Create a separate channel for each area of high load** in the
+        application, or
+
+    2.  **Use a pool of gRPC channels** to randomly distribute RPCs over
+        multiple connections (channels must have different channel args to
+        prevent re-use so define a use-specific channel arg such as channel
+        number). \
+        ***Side note:*** *The gRPC team has plans to add a feature to fix these
+        performance issues, so any solution involving creating multiple channels
+        is a temporary workaround that should eventually not be needed.*
+
+### C++
+
+*   **Do not use Sync API for performance sensitive servers.** If performance
+    and/or resource consumption are not concerns, use the Sync API as it is the
+    simplest to implement for low-QPS services.
+
+*   **Favor callback API over other APIs for all RPCs**, given that the
+    application can avoid all blocking operations or blocking operations can be
+    moved to a separate thread.
+
+*   If having to use the async completion-queue API, the **best scalability
+    trade-off is having numcpu’s threads and one completion queue per thread**.
+
+*   For the async completion-queue API, make sure to **register enough server
+    requests for the desired level of concurrency** to avoid the server
+    continuously getting stuck in a slow path that results in essentially serial
+    request processing. 
+
+*   **Enable write batching in streams** if message k + 1 does not rely on
+    responses from message k by passing a WriteOptions argument to Write with
+    buffer_hint set - `stream_writer->Write(message,
+    WriteOptions().set_buffer_hint());`
+
+*   *(Special topic)* In the case that your application has unary or server-side
+    streaming methods where there will be exactly one request message from the
+    client, the **synchronous-with-deferred-read interface** could reduce memory
+    usage on busy servers. Read more about the interface
+    [here](https://g3doc.corp.google.com/net/grpc/g3doc/cpp/codegen.md#synchronous-with-deferred-read-interface).
+
+*   *(Special topic)*
+    [gRPC::GenericStub](https://grpc.github.io/grpc/cpp/grpcpp_2generic_2generic__stub_8h.html)
+    can be useful in certain cases when there is high contention / CPU time
+    spent on proto serialization. This class allows the application to directly
+    send **raw gRPC::ByteBuffer as data** rather than serializing from some
+    proto. This can also be helpful if the same data is being sent multiple
+    times, with one explicit proto-to-ByteBuffer serialization followed by
+    multiple ByteBuffer sends. 
+
+### Java
+
+*   **Use non-blocking stubs** to parallelize RPCs.
+
+*   **Provide a custom executor based on the workload** (cached (default), fixed, forkjoin, etc).
```

Comment:
The main thing here is that the executor would have a limit on number of threads to avoid excessive context switches and memory overhead under high load. So I'd say, "Provide a custom executor that limits the number of threads, based on your workload."